datalad_dataverse/remote.py

```diff
@@ -92,111 +136,311 @@ def api(self):
+            #   'restricted': False,
+            #   'directoryLabel': 'subdir2',
+            #   'version': 1,
+            #   'datasetVersionId': 72,
+            #   'dataFile': {'id': 682,
+            #   'persistentId': '',
+            #   'pidURL': '',
+            #   'filename': 'third_file.md',
+            #   'contentType': 'text/plain',
+            #   'filesize': 9,
+            #   'description': '',
+            #   'storageIdentifier': 'local://1821bc70e68-c3c9dedcfce6',
+            #   'rootDataFileId': -1,
+            #   'md5': 'd8d77109f4a24efc3bd53d7cabb7ee35',
+            #   'checksum': {'type': 'MD5',
+            #                'value': 'd8d77109f4a24efc3bd53d7cabb7ee35'},
+            #   'creationDate': '2022-07-20'}
+
+            # sort by version, so we can rely on first entry to refer to latest
+            # Note, that ('versionNumber', 'versionMinorNumber', 'versionState')
+            # would look like this:
+            # (None, None, 'DRAFT'), (2, 0, 'RELEASED'), (1, 0, 'RELEASED')
+            # and we need a possible DRAFT to have the biggest key WRT sorting.
+            self._dataset.sort(key=lambda v:
+                                   (v.get('versionNumber') or sys.maxsize,
+                                    v.get('versionMinorNumber') or sys.maxsize),
+                               reverse=True)
+
+        return self._dataset
+
+    @property
+    def files_all(self):
+        """Cache of files available from dataverse.
+
+        For quick lookup and deduplication, this is a dict {id: remote_path}
+        where remote_path is a `Path` build from the file's 'directoryLabel'
+        (if any) and 'filename'.
+        Currently, this is all we need, hence don't bother carrying and updating
+        all other metadata.
+        """
+
+        if self._files_all is None:
+            self._files_all = {
+                f['dataFile']['id']: Path(f.get('directoryLabel', '')) / f['dataFile']['filename']
+                for f_l in [version['files'] for version in self.dataset]
+                for f in f_l}
+
+        return self._files_all
+
+    @property
+    def files_latest(self):
+        """Cache of files in the latest version of the dataverse dataset.
+
+        This refers to the DRAFT version (if there is any) or the latest
+        published version otherwise. That's the version pushes go into. Hence,
+        this is needed to determine whether we need to replace a file on
+        upload, while the complete list in `self.files_all` is relevant for key
+        retrieval (which can come from earlier dataverse versions of the
+        dataset).
+        """
+
+        if self._files_latest is None:
+            # Latest version in self.dataset is first entry.
+            self._files_latest = {
+                f['dataFile']['id']: Path(f.get('directoryLabel', '')) / f['dataFile']['filename']
+                for f in self.dataset[0]['files']}
```

Comment:
So for any unpublished dataset, `files_all` and `files_latest` are identical.

it would be good if both would use the same code.